### PR TITLE
[FLINK-23926] one timestamp

### PR DESCRIPTION
**Pull request by alpinegizmo**
This PR is layered on top of https://github.com/apache/flink-training/pull/31.

The objective is to replace the startTime and endTime fields with a single eventTime field. This is more natural, and will make it more straightforward to convert a `DataStream<TaxiRide>` into a table.

This does affect the long rides exercise. The biggest change is that it is no longer possible to determine if a ride has been too long in the case where the START event is missing.

### CODE REVIEW

File: common/src/main/java/org/apache/flink/training/exercises/common/datatypes/TaxiRide.java
Comment (NicoK) [inline common/src/main/java/org/apache/flink/training/exercises/common/datatypes/TaxiRide.java]: Why not use `java.time.Instant#compareTo`?
```suggestion
        int compareTimes = this.eventTime.compareTo(other.eventTime);
```

File: common/src/main/java/org/apache/flink/training/exercises/common/sources/TaxiRideGenerator.java
Comment (NicoK) [inline common/src/main/java/org/apache/flink/training/exercises/common/sources/TaxiRideGenerator.java]: Not for this PR, but didn't you want to remove `collectWithTimestamp` calls?

If I see this correctly, only the LongRides exercises and solutions need timestamps and they have appropriate watermark strategies...

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (NicoK) [inline long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: Can you add a section to the discussion on how to deal with missing START events? We could use event-time timers for this as well, e.g. by firing for the END event's timestamp

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (NicoK) [inline long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java]: There **is** a timer if `firstRideEvent` was a START event...

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (NicoK) [inline long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java]: You keep removing "yet" here...isn't this making clear that the timer would have fired if we hadn't removed it above?
```suggestion
                    // it could be that the ride has gone on too long (and the timer didn't fire yet)
```

File: long-ride-alerts/src/solution/scala/org/apache/flink/training/solutions/longrides/scala/LongRidesSolution.scala
Comment (NicoK) [inline long-ride-alerts/src/solution/scala/org/apache/flink/training/solutions/longrides/scala/LongRidesSolution.scala]:
```suggestion
          // it could be that the ride has gone on too long (and the timer didn't fire yet)
```

File: long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesUnitTest.java
Comment (NicoK) [inline long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesUnitTest.java]: From the Javadoc:
> You must set it before calling the assertion otherwise it is ignored as the failing assertion breaks the chained call by throwing an AssertionError.

```suggestion
        assertThat(results.size())
                .withFailMessage("Expecting test to have exactly one result")
                .isEqualTo(1);
```

File: long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesUnitTest.java
Comment (NicoK) [inline long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesUnitTest.java]: Converting to array is unnecessary:
```suggestion
        StreamRecord<Long> resultingRecord = (StreamRecord<Long>) results.element();
```

**Review by NicoK — CHANGES_REQUESTED** [review]
Thanks for getting this change in - it maps better to the streaming approach.

I have a couple of smaller things around the code but once addressed, we can merge

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (alpinegizmo) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: It's not possible to handle missing START events. Those START events are now the only source of knowledge about when the ride started, so it's not possible to know how long the ride was.

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (alpinegizmo) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java]: I've reworked the comments

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (NicoK) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: True, you won't know how long a ride was, but you can identify a late START event at least (and maybe put it on a side output for error handling - a best practice in these cases but not sure you want to add this here

Comment (alpinegizmo) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: There's no notion of lateness here. The current solution will wait indefinitely for the START event. If it ever shows up, the ride will then be reported as having been too long.

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (alpinegizmo) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java]: reworked, as above

File: long-ride-alerts/src/solution/scala/org/apache/flink/training/solutions/longrides/scala/LongRidesSolution.scala
Comment (alpinegizmo) [reply long-ride-alerts/src/solution/scala/org/apache/flink/training/solutions/longrides/scala/LongRidesSolution.scala]: fixed

File: long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
Comment (alpinegizmo) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: I guess I see what you meant. We could detect missing START events, and indicate that on a side output. Maybe I'll add something to the DISCUSSION about that.

Comment (NicoK) [reply long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java:127]: Exactly, this is what I meant

File: long-ride-alerts/DISCUSSION.md
Comment (NicoK) [inline long-ride-alerts/DISCUSSION.md]: That sounds wrong...did you mean this?
```suggestion
*These solutions also leak state whenever there is a long ride and the END event is missing.*
```

**Review by NicoK — APPROVED** [review]
One tiny bit in a sentence seems wrong. I'll approve now assuming you'll address that one

File: long-ride-alerts/DISCUSSION.md
Comment (alpinegizmo) [reply long-ride-alerts/DISCUSSION.md]: No, that isn't what I meant. I've expanded the DISCUSSION to be clearer.

Comment (NicoK) [reply long-ride-alerts/DISCUSSION.md]: The sentence is still misleading as it suggests that you always leak state for every long ride but this is only true if the start event is missing or if the end event is late. Long rides with both events are cleaned up in `processElement`.
The text should thus go through these (or reflect them somehow):
- start event missing -> end event sits in state (leak!)
- end event missing -> timer fires and clears state (ok)
- end event late -> timer fires, clears state, end event arrives and is stored in state (leak!)

**Review by NicoK — CHANGES_REQUESTED** [review]

**Comment by alpinegizmo** [comment]
@NicoK I've reworked the DISCUSSION for the long-rides exercise again. It should be correct now.

**Review by NicoK — APPROVED** [review]

**Pull request merged**
